src/frontend/src/scheduler/local.rs

```diff
@@ -453,4 +454,39 @@ impl LocalQueryExecution {
             }
         }
     }
+
+    #[inline(always)]
+    fn get_vnode_mapping(&self, table_id: &TableId) -> Option<ParallelUnitMapping> {
+        self.front_env
+            .catalog_reader()
+            .read_guard()
+            .get_table_by_id(table_id)
+            .map(|table| {
+                self.front_env
+                    .worker_node_manager()
+                    .get_fragment_mapping(&table.fragment_id)
+            })
+            .ok()
+            .flatten()
+    }
+
+    fn choose_worker(&self, stage: &Arc<QueryStage>) -> SchedulerResult<Vec<WorkerNode>> {
+        if stage.parallelism.unwrap() == 1 {
+            if let NodeBody::Insert(insert_node) = &stage.root.node
+                && let Some(vnode_mapping) = self.get_vnode_mapping(&insert_node.table_id.into()) {
+                    let worker_node = {
+                        let parallel_unit_ids = vnode_mapping.iter_unique().collect_vec();
+                        let candidates = self.front_env
+                            .worker_node_manager()
+                            .get_workers_by_parallel_unit_ids(&parallel_unit_ids)?;
+                        candidates.choose(&mut rand::thread_rng()).unwrap().clone()
+                    };
```

Comment:
Why do we need `get_workers_by_parallel_unit_ids` here? Didn't use this before, so not too familiar.
In the case of `Insert` why is this needed?

Reply:
The insert executor will send the insert data to the reader registered by the dml executor.
https://github.com/risingwavelabs/risingwave/blob/1ad23babf8b61e89cbe9ffc3a8fadb7f878f9ffe/src/stream/src/executor/dml.rs#L94
Hence to access the reader, the insert executor need to schedule the same worker node of the dml executor. So `get_workers_by_parallel_unit_ids` is to get the worker node where dml executor stay in.

Reply:
I see, thanks for the clear explanation!

Maybe this can be documented, since the logic is split in various places, it does not seem very clear to me at first glance.
(Unless some documentation already exists, in that case feel free to ignore).

Reply:
Agree with you. (Seems don't have a related doc)